configs/example/gem5_library/multisim/example_callable.py

```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2024 The Regents of the University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from gem5.isas import ISA
+from gem5.prebuilt.riscvmatched.riscvmatched_board import RISCVMatchedBoard
+from gem5.simulate.simulator import Simulator
+from gem5.utils.requires import requires
+
+
+def run_riscvmathed_workload(resource) -> Simulator:
+    requires(isa_required=ISA.RISCV)
+
+    # instantiate the riscv matched board with default parameters
+    board = RISCVMatchedBoard()
+
+    board.set_workload(resource)
+
+    # run the simulation with the RISCV Matched board
+    simulator = Simulator(board=board, full_system=False)
+    return simulator
+
+
+def run_riscvmatched_worklaod_diff_clocks(resource) -> Simulator:
+    requires(isa_required=ISA.RISCV)
+
+    # instantiate the riscv matched board with default parameters
+    board = RISCVMatchedBoard(
+        clk_freq="2.2GHz",
+    )
+
+    board.set_workload(resource)
+
+    # run the simulation with the RISCV Matched board
+    simulator = Simulator(board=board, full_system=False)
+    return simulator
```

Comment:
This is all a bit clunky. I'd suggest returning partial functions. In this example you're returning Simulator _objects_, not a function which can be called. I'd suggest a change like this (haven't actually ran this but I'm pretty sure it's the right way to do things):

```suggestion
from gem5.isas import ISA
from gem5.prebuilt.riscvmatched.riscvmatched_board import RISCVMatchedBoard
from gem5.simulate.simulator import Simulator
from gem5.utils.requires import requires
from typing import List, Callable, Collection
from functools import partial

requires(isa_required=ISA.RISCV)

def get_simulator_partials(workloads: Collection[WorkloadResource]) -> List[Callable[[], Simulator]:
    to_return = []
    for clk_freq in ["2.2GHz", "3.6GHz", "4GHz"]:
        for workload in workloads:
            board = RISCVMatchedBoard(clk_freq=clk_freq)
            board.set_workload(workload)
            to_return.append(partial(Simulator, board))
    return to_return
```

Here we have a function which takes in a collection of workloads. For each workload the a simulation with the RISCV Board is setup 3 times, each with a different clock frequency.

Reply:
If this works +1.

`to_return.append(lambda: Simulator(board))` may work as well if you don't want to use `partial`. I can't decide which I like more.